fem/nonlinearform.cpp

```diff
@@ -320,26 +473,42 @@ Operator &NonlinearForm::GetGradient(const Vector &x) const
       *Grad = 0.0;
    }
 
-   if (dnfi.Size())
+   if (domain_integs.Size())
    {
+      for (int k = 0; k < domain_integs.Size(); k++)
+      {
+         if (domain_integs_marker[k] != nullptr)
+         {
+            MFEM_VERIFY(mesh->attributes.Size() ==
+                        domain_integs_marker[k]->Size(),
+                        "invalid element marker for domain integrator #"
+                        << k << ", counting from zero");
+         }
+      }
+
       for (int i = 0; i < fes->GetNE(); i++)
       {
+         int elem_attr = fes->GetMesh()->GetAttribute(i);
          fe = fes->GetFE(i);
          doftrans = fes->GetElementVDofs(i, vdofs);
          T = fes->GetElementTransformation(i);
          px.GetSubVector(vdofs, el_x);
          if (doftrans) {doftrans->InvTransformPrimal(el_x); }
-         for (int k = 0; k < dnfi.Size(); k++)
+         for (int k = 0; k < domain_integs.Size(); k++)
          {
-            dnfi[k]->AssembleElementGrad(*fe, *T, el_x, elmat);
-            if (doftrans) { doftrans->TransformDual(elmat); }
-            Grad->AddSubMatrix(vdofs, vdofs, elmat, skip_zeros);
-            // Grad->AddSubMatrix(vdofs, vdofs, elmat, 1);
+            if (domain_integs_marker[k] == nullptr ||
+                (*(domain_integs_marker[k]))[elem_attr-1] == 1)
+            {
+               domain_integs[k]->AssembleElementGrad(*fe, *T, el_x, elmat);
+               if (doftrans) { doftrans->TransformDual(elmat); }
+               Grad->AddSubMatrix(vdofs, vdofs, elmat, skip_zeros);
+               // Grad->AddSubMatrix(vdofs, vdofs, elmat, 1);
```

Comment:
Can this be removed?

Reply:
Do you mean the if that checks `(domain_integs_marker[k] == nullptr || (*(domain_integs_marker[k]))[elem_attr-1] == 1)`? I copied that over from the bilinear form assembly routine, and since I added the `domain_integs_marker` to the nonlinear form it is required. I added the `domain_integs_marker` to make the interface of the nonlinear and bilinear forms more consistent, and because I wanted to be able to use it for calls to `GetEnergy`.